pageserver/src/layered_repository.rs

```diff
@@ -0,0 +1,1056 @@
+    }
+
+    fn get_next_tag(
+        &self,
+        _tag: crate::object_key::ObjectTag,
+    ) -> Result<Option<crate::object_key::ObjectTag>> {
+        todo!();
+    }
+
+    ///
+    /// Flush to disk all data that was written with the put_* functions
+    ///
+    /// NOTE: This has nothing to do with checkpoint in PostgreSQL. We don't
+    /// know anything about them here in the repository.
+    fn checkpoint(&self) -> Result<()> {
+
+        let last_valid_lsn = self.last_valid_lsn.load();
+        trace!("checkpointing timeline {} at {}", self.timelineid, last_valid_lsn);
+
+        let mut layers = self.layers.lock().unwrap();
+
+        // Walk through each in-memory, and write any dirty data to disk,
+        // as snapshot files.
+        //
+        // We currently write a new snapshot file for every relation
+        // that was modified, if there has been any changes at all.
+        // It would be smarter to only flush out in-memory layers that
+        // have accumulated a fair amount of changes. Note that the
+        // start and end LSNs of snapshot files belonging to different
+        // relations don't have to line up, although currently they do
+        // because of the way this works. So you could have a snapshot
+        // file covering LSN range 100-200 for one relation, and a
+        // snapshot file covering 150-250 for another relation. The
+        // read functions should even cope with snapshot files
+        // covering overlapping ranges for the same relation, although
+        // that situation never arises currently.
+        //
+        // Note: We release all the layer structs, and start fresh
+        // with an empty map. This keeps memory usage in check, but is
+        // probably too aggressive. Some kind of LRU policy would be
+        // appropriate.
+        //
+
+        // Call freeze() for any unfrozen layers (that is, layers that
+        // haven't been written to disk yet)
+        // Call unload() for all layers, to release memory.
+        //
+        // FIXME: We do this by creating a whole new LayerMap. I couldn't
+        // figure out the borrowing rules to remove and insert entries
+        // to the old LayerMap while iterating through it.
+        let old_layers = std::mem::take(&mut *layers);
+        for layer in old_layers.inner.values() {
+            if !layer.is_frozen() {
+                let new_layers = layer.freeze(last_valid_lsn, &*self.walredo_mgr)?;
+
+                for new_layer in new_layers {
+                    info!("freeze returned {} {}-{}", new_layer.get_relish_tag(), new_layer.get_start_lsn(), new_layer.get_end_lsn());
+                    layers.insert(Arc::clone(&new_layer));
+                }
+            } else {
+                // FIXME: if this fails, we're in trouble because we already
+                // swapped the 'layers' with the empty one. Hence panic on error.
+                layer.unload().expect("could not unload layer from memory");
```

Comment:
It's actually harmless if the 'unload' fails, this could be just a warning to the log instead. However returning with `?` would be bad, because we had already replaced 'layers' with an empty map. It's OK to 'take' the layer map first, because we're holding the mutex on it, but if we do that, we must not return from the function until we're replaced it with `new_layers`. Now that I look at this again, I see that the `layer.freeze(..)?` has the same problem; if that fails, we are left with an empty layers map.

I'll try to refactor that to make it more safe.